src/main/java/com/epam/aidial/core/upstream/UpstreamRouteProvider.java

```diff
@@ -0,0 +1,122 @@
+     */
+    public UpstreamRoute get(UpstreamProvider provider) {
+        String deploymentName = provider.getName();
+        List<Upstream> upstreams = provider.getUpstreams();
+
+        TieredBalancer balancer = balancers.get(deploymentName);
+        if (balancer == null) {
+            // if no state found for upstream, it's probably custom application
+            balancer = new TieredBalancer(deploymentName, upstreams);
+        }
+
+        return new UpstreamRoute(balancer, MAX_RETRY_COUNT);
+    }
+
+    public synchronized void onUpdate(Config config) {
+        log.debug("Updating load balancers state");
+        Map<String, TieredBalancer> oldState = balancers;
+        Map<String, TieredBalancer> newState = new HashMap<>();
+        Set<String> uniqueDeployments = new HashSet<>();
+
+        Map<String, Model> models = config.getModels();
+        updateDeployments(newState, oldState, uniqueDeployments, models.values());
+
+        Map<String, Application> applications = config.getApplications();
+        updateDeployments(newState, oldState, uniqueDeployments, applications.values());
+
+        Map<String, Addon> addons = config.getAddons();
+        updateDeployments(newState, oldState, uniqueDeployments, addons.values());
+
+        Map<String, Assistant> assistants = config.getAssistant().getAssistants();
+        updateDeployments(newState, oldState, uniqueDeployments, assistants.values());
+
+        LinkedHashMap<String, Route> routes = config.getRoutes();
+        updateRoutes(newState, oldState, uniqueDeployments, routes.values());
+
+        balancers = newState;
+    }
+
+    private static void updateRoutes(Map<String, TieredBalancer> newState, Map<String, TieredBalancer> oldState,
+                                     Set<String> uniqueDeployments, Collection<Route> routes) {
+        for (Route route : routes) {
+            String name = route.getName();
+
+            RouteEndpointProvider endpointProvider = new RouteEndpointProvider(route);
+            TieredBalancer previous = oldState.get(name);
+            updateDeployment(endpointProvider, previous, uniqueDeployments, newState);
+        }
+    }
+
+    private static void updateDeployments(Map<String, TieredBalancer> newState, Map<String, TieredBalancer> oldState,
+                                          Set<String> uniqueDeployments, Collection<? extends Deployment> deployments) {
+        for (Deployment deployment : deployments) {
+            String name = deployment.getName();
+
+            DeploymentUpstreamProvider endpointProvider = new DeploymentUpstreamProvider(deployment);
+            TieredBalancer previous = oldState.get(name);
+            updateDeployment(endpointProvider, previous, uniqueDeployments, newState);
+        }
+    }
+
+    private static void updateDeployment(UpstreamProvider upstream, TieredBalancer previous, Set<String> uniqueDeployments,
+                                         Map<String, TieredBalancer> newState) {
```

Comment:
Why you can't just use Map<String, TieredBalancer> newState to check for duplicates?

Reply:
I see Long.parseLong() which can return negatives.